### PR TITLE
[Bugfix] : Machine controllers wrong display/config ignored

### PR DESCRIPTION
**Pull request by galyfray**
**What:**
Fixed Machine controller slide bar not showing correct bounds
Fixed Machine controller setting being ignored if multiple controller control the same machine.

**How solved:**
Set maximum of `SliderWidge` to 15 and minimum to 0.
Set value loaded from NBT to 15 if it is above it.
Added two methods `shouldAllowWorking()` which is just moved code from `updateRedstoneStatus` and `isOtherAllowingWork()` for checking if other controllers will allow the machine to work.

**Outcome:**
Fixes : #1161 
Fixes : #1268 

**Additional info:**
Machine controller ui

**Possible compatibility issue:**
Possible compatibility issue with older setup if the controller was set to redstone level 16, briefly tested with other level no compatibility issue found.

### CODE REVIEW

**Comment by LAGIdiot** [comment]
Updated PR massage to correctly trigger linking of issues

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 1]
```diff
@@ -182,17 +183,30 @@ private IControllable getControllable() {
...
+
+    private boolean isOtherAllowingWork(){
+        boolean otherAllow = true;
+        for (EnumFacing side : EnumFacing.values()){
+            if(side != attachedSide && coverHolder.getCoverAtSide(side) instanceof  CoverMachineController ){
+                otherAllow = otherAllow && ((CoverMachineController)coverHolder.getCoverAtSide(side)).shouldAllowWorking();
+            }
         }
+        return otherAllow;
     }
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:210]: You need also to check in which mode does controller work as they may not be controlling same thing. For example one is controlling machine and other one Pump Cover.

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 2]
```diff
@@ -182,17 +183,30 @@ private IControllable getControllable() {
...
         IControllable controllable = getControllable();
         if(controllable != null) {
-            controllable.setWorkingEnabled(shouldAllowWorking);
+            controllable.setWorkingEnabled(shouldAllowWorking() && isOtherAllowingWork());
+        }
+    }
+
+    private boolean shouldAllowWorking(){
+        boolean shouldAllowWorking = getRedstoneSignalInput() < minRedstoneStrength;
+        return  isInverted != shouldAllowWorking ;// equivalent to isInverted ? !shouldAllowWorking: shouldAllowWorking
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:199]: This constructs is way too hard to digest. When there must be comment to explain boolean logic than that logic is over-complicated and may lead to errors. I know IDE suggested but I won't allow it.

```suggestion
        //noinspection SimplifiableConditionalExpression
        return isInverted ? !shouldAllowWorking : shouldAllowWorking;
```

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 3]
```diff
@@ -182,17 +183,30 @@ private IControllable getControllable() {
...
+    }
+
+    private boolean isOtherAllowingWork(){
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:202]: Naming of this function does not sound right to me, maybe better naming would be `doesOtherAllowWorking`

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 4]
```diff
@@ -182,17 +183,30 @@ private IControllable getControllable() {
...
-        if(isInverted) shouldAllowWorking = !shouldAllowWorking;
         IControllable controllable = getControllable();
         if(controllable != null) {
-            controllable.setWorkingEnabled(shouldAllowWorking);
+            controllable.setWorkingEnabled(shouldAllowWorking() && isOtherAllowingWork());
+        }
+    }
+
+    private boolean shouldAllowWorking(){
+        boolean shouldAllowWorking = getRedstoneSignalInput() < minRedstoneStrength;
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:198]: This does not work correctly if both `minRedstoneStrength` and `getRedstoneSignalInput()` are set to 15. But I am not sure how to handle this yet. I may debate this on Discord.

**Review by LAGIdiot — CHANGES_REQUESTED** [review]
Thanks for creating this PR. I have found some areas that requires some adjustments so please see to them.

File: src/main/java/gregtech/common/covers/CoverMachineController.java (hunk 4 above)
Comment (galyfray) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:198]: if both are set to 15 then the controller don't work if we use `<=` instead of `<` then the machine will always work even when directly putting a lever on it.

File: src/main/java/gregtech/common/covers/CoverMachineController.java (hunk 3 above)
Comment (galyfray) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:202]: yep wasn't nice for me too I will replace

**Comment by galyfray** [comment]
changes done thank you for your feedback !

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 5]
```diff
@@ -209,6 +223,7 @@ public void readFromNBT(NBTTagCompound tagCompound) {
         this.minRedstoneStrength = tagCompound.getInteger("MinRedstoneStrength");
         this.isInverted = tagCompound.getBoolean("Inverted");
         this.controllerMode = ControllerMode.values()[tagCompound.getInteger("ControllerMode")];
+        if (minRedstoneStrength > 15) minRedstoneStrength = 15;
```
[the remark this answers is not in the document]
Comment (LAGIdiot) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:228]: This one does not seems resolved to me and I still require this change.

File: src/main/java/gregtech/common/covers/CoverMachineController.java (hunk 4 above)
Comment (LAGIdiot) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:197]: That would be correct statement but in this case it would lower requirements for Redstone signal by one level. Let's say from 8 to 7 which may break someone builds. And we always try to maintain high level of backward compatibility (not always possible). So in this case we need to add some special logic which will handle 15 case (and properly documented it - yes we have many hacks without documentation but I won't all any new ones missing documentation during my reign) and let's everything else work as before.

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 6]
```diff
@@ -115,7 +115,7 @@ public ModularUI createUI(EntityPlayer player) {
         updateDisplayInventory();
         return ModularUI.defaultBuilder()
             .label(10, 5, "cover.machine_controller.name")
-            .widget(new SliderWidget("cover.machine_controller.redstone", 10, 20, 156, 20, 1.0f, 16.0f,
+            .widget(new SliderWidget("cover.machine_controller.redstone", 10, 20, 156, 20, 1f, 15.0f,
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:118]: There is unnecessary change regarding 1.0f -> 1f it will not have any effect on code but for consistency sake and for making PRs as small as possible (so it's more readable).

**Review by LAGIdiot — CHANGES_REQUESTED** [review]
Thanks for your hard work on this PR. I have added some comments to specific section that are not up to speed yet. And I have one overall comment, some sections of code seems not properly formatted so please run IDE formatter on all files that you changed before committing them.

I may be requiring quite a lot from you but I need to maintain code quality, readability and maintainability.

File: src/main/java/gregtech/common/covers/CoverMachineController.java (hunk 6 above)
Comment (galyfray) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:118]: as you can see here :
 https://github.com/GregTechCE/GregTech/blob/59cf924550f97635e1f576bf68309adcb8bface9/src/main/java/gregtech/common/covers/CoverMachineController.java#L190
my PR should not make any breaking change by reducing redstone requirement.
I understand that maintaining overall code quality is important don't worry I'll did any required changes until it become perfect.

File: src/main/java/gregtech/common/covers/CoverMachineController.java (hunk 4 above)
Comment (galyfray) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:197]: as you can see here :
 https://github.com/GregTechCE/GregTech/blob/59cf924550f97635e1f576bf68309adcb8bface9/src/main/java/gregtech/common/covers/CoverMachineController.java#L190
my PR should not make any breaking change by reducing redstone requirement.
I understand that maintaining overall code quality is important don't worry I'll did any required changes until it become perfect.

Comment (LAGIdiot) [reply src/main/java/gregtech/common/covers/CoverMachineController.java:197]: Yeah you are right I had that logic inverted in my head because this concept prevents working by default not allows it. My bad. This is correct implementation.

File: src/main/java/gregtech/common/covers/CoverMachineController.java [hunk 7]
```diff
@@ -207,6 +223,7 @@ public void writeToNBT(NBTTagCompound tagCompound) {
     public void readFromNBT(NBTTagCompound tagCompound) {
         super.readFromNBT(tagCompound);
         this.minRedstoneStrength = tagCompound.getInteger("MinRedstoneStrength");
+        if (minRedstoneStrength > 15) minRedstoneStrength = 15;
```
Comment (LAGIdiot) [inline src/main/java/gregtech/common/covers/CoverMachineController.java:226]: Missing this...
```suggestion
        if (minRedstoneStrength > 15) this.minRedstoneStrength = 15;
```

**Review by LAGIdiot — APPROVED** [review]
Thanks for your hard work on this PR. I have added two comments and will merge it with small fix,
